### ISSUE TITLE
many-to-many relations with oracle

### ISSUE THREAD

**Issue by onlygecko** [kvesteri/sqlalchemy-continuum#91]
Hello,

I'm trying to use sqlalchemy-continuum with oracle. I had to manipulate the Transaction class to make it work with Oracle. Oracle uses Sequences and is not working with autoincrement (http://docs.sqlalchemy.org/en/rel_0_9/dialects/oracle.html?highlight=oracle#auto-increment-behavior).

Thanks to sqlalchemy-continuum I'm able to version single models and one-to-many relationships but I have problems versioning many-to-many relationships. 

``` python
# -*- coding: utf-8 -*-

import sqlalchemy as sa
from sqlalchemy.orm import relationship
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy.orm import sessionmaker
from sqlalchemy import create_engine
from sqlalchemy_continuum import make_versioned

Base = declarative_base()
make_versioned(user_cls=None)

#------------- Models -------------------------------------#

association_table = sa.Table('associations', Base.metadata,
    sa.Column('article_id', sa.Integer, sa.ForeignKey('articles.id'),primary_key=True),
    sa.Column('tag_id', sa.Integer, sa.ForeignKey('tags.id'),primary_key=True),
)

class Article(Base):

    __tablename__ = 'articles'
    __versioned__ = {}
    id = sa.Column(sa.Integer, sa.Sequence('article_seq'), primary_key=True)
    name = sa.Column(sa.Unicode(255), nullable=False)
    content = sa.Column(sa.UnicodeText)

    def __init__(self, name, content):
        self.name = name
        self.content = content


class Tag(Base):

    __tablename__ = 'tags'
    __versioned__ = {}
    id = sa.Column(sa.Integer, sa.Sequence('tag_seq'), primary_key=True)
    name = sa.Column(sa.Unicode(255), nullable=False)
    articles = relationship(Article, secondary=association_table, backref="tags")

    def __init__(self,name):
        self.name = name

    def __str__(self):
        return self.name

#------------- DB-Connection -------------------------------------------#

dialect='oracle'
db_server = 'myserver:1521'
db_name = 'mydb'
db_user = 'user'
db_password = 'pw'
engine = create_engine('%s://%s:%s@%s/%s' % (dialect, db_user, db_password, db_server, db_name),
                       pool_recycle=300, pool_size=20, max_overflow=40)

sa.orm.configure_mappers()
Base.metadata.create_all(bind=engine)
Session = sessionmaker(bind=engine)
session = Session()


#------------- Example-Data -------------------------------------------#
tag_it = Tag('IT')
session.add(tag_it)
article_one = Article('RaspberriPy 2', 'the new raspberry pi received')
article_one.tags.append(tag_it)
session.add(article_one)
session.commit()
```

The example above results in the following error:

```
  File "/home/gecko/workspace/sa-continuum-test/run_many_to_many.py", line 69, in <module>
    session.commit()
  File "/home/gecko/opt/pyenv/sa-continuum/local/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 776, in commit
    self.transaction.commit()
  File "/home/gecko/opt/pyenv/sa-continuum/local/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 377, in commit
    self._prepare_impl()
  File "/home/gecko/opt/pyenv/sa-continuum/local/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 357, in _prepare_impl
    self.session.flush()
  File "/home/gecko/opt/pyenv/sa-continuum/local/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 1919, in flush
    self._flush(objects)
  File "/home/gecko/opt/pyenv/sa-continuum/local/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 2037, in _flush
    transaction.rollback(_capture_exception=True)
  File "/home/gecko/opt/pyenv/sa-continuum/local/lib/python2.7/site-packages/sqlalchemy/util/langhelpers.py", line 60, in __exit__
    compat.reraise(exc_type, exc_value, exc_tb)
  File "/home/gecko/opt/pyenv/sa-continuum/local/lib/python2.7/site-packages/sqlalchemy/orm/session.py", line 2001, in _flush
    flush_context.execute()
  File "/home/gecko/opt/pyenv/sa-continuum/local/lib/python2.7/site-packages/sqlalchemy/orm/unitofwork.py", line 372, in execute
    rec.execute(self)
  File "/home/gecko/opt/pyenv/sa-continuum/local/lib/python2.7/site-packages/sqlalchemy/orm/unitofwork.py", line 481, in execute
    self.dependency_processor.process_saves(uow, states)
  File "/home/gecko/opt/pyenv/sa-continuum/local/lib/python2.7/site-packages/sqlalchemy/orm/dependency.py", line 1090, in process_saves
    secondary_update, secondary_delete)
  File "/home/gecko/opt/pyenv/sa-continuum/local/lib/python2.7/site-packages/sqlalchemy/orm/dependency.py", line 1134, in _run_crud
    connection.execute(statement, secondary_insert)
  File "/home/gecko/opt/pyenv/sa-continuum/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 729, in execute
    return meth(self, multiparams, params)
  File "/home/gecko/opt/pyenv/sa-continuum/local/lib/python2.7/site-packages/sqlalchemy/sql/elements.py", line 322, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/home/gecko/opt/pyenv/sa-continuum/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 826, in _execute_clauseelement
    compiled_sql, distilled_params
  File "/home/gecko/opt/pyenv/sa-continuum/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 958, in _execute_context
    context)
  File "/home/gecko/opt/pyenv/sa-continuum/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1159, in _handle_dbapi_exception
    exc_info
  File "/home/gecko/opt/pyenv/sa-continuum/local/lib/python2.7/site-packages/sqlalchemy/util/compat.py", line 199, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb)
  File "/home/gecko/opt/pyenv/sa-continuum/local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 951, in _execute_context
    context)
  File "/home/gecko/opt/pyenv/sa-continuum/local/lib/python2.7/site-packages/sqlalchemy/engine/default.py", line 436, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.DatabaseError: (DatabaseError) ORA-01036: illegal variable name/number
 'INSERT INTO associations (article_id, tag_id) VALUES (:article_id, :tag_id)' {'operation_type': 0, 'article_id': 4, 'tag_id': 6}
```

Do you now, why SQLAlchemy tries to insert values for columns which does not exist?

**Comment by kvesteri** [comment]
Hmm very strange. It seems it tries to feed the associations_version table values to association table. Its hard for me to reproduce this since I don't have access to Oracle database :/

**Comment by onlygecko** [comment]
Could you tell me, where I should look or how could get to the bottom of this problem by myself?

**Comment by kvesteri** [comment]
I would look probably start looking for the code that generates association versions: https://github.com/kvesteri/sqlalchemy-continuum/blob/master/sqlalchemy_continuum/manager.py#L377

**Comment by onlygecko** [comment]
Thanks for the tip. It was necessary to pass a copy of the insert params instead of the dict it self. Otherwise the dict get's a new parameter and which leads to an error on insertation. (https://github.com/onlygecko/sqlalchemy-continuum/commit/af157c7e2d2adddda9b9434a299e78baa28b6126)

Is this a good solution or do I miss something?

**Comment by fabioramponi** [comment]
Hi guys, 
I'm having the same issue with many-to-many relationships versioning in Oracle. Any update on this?

**Comment by AbdealiJK** [comment]
Just realized I faced the same issue and have created a duplicate ticket at https://github.com/kvesteri/sqlalchemy-continuum/issues/289
Have also provided a pull request which fixed the issue at: https://github.com/kvesteri/sqlalchemy-continuum/pull/291

**Comment by marksteward** [comment]
Fixed by #291
